Patch add str begins and ends

wkhudgins92 opened the pull request:
This is the patch for RFC
https://wiki.php.net/rfc/add_str_begin_and_end_functions

This PR contains the implementations of str_begins() and str_ends() and also the phpt tests.

wkhudgins92 commented:
Not sure what the build failure is about. I was able to compile this on my local machine using `make -j2` and I was able to run the compiled PHP and also execute `make test`.

nikic commented on ext/standard/string.c:
```diff
@@ -2445,6 +2445,46 @@ PHP_FUNCTION(substr)
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "SS|l", &str, &search_value, &case_sensitive) == FAILURE) 
+		RETURN_NULL();
+
+	for (int i = 0; i < search_value->len; i++) {
```
The `int i` should go to the top of the function. That should fix the build issue.

laruence commented:
how does str_begins different from strncmp/strncasecmp?

````
<?php
$str = "bEginningMiddleEnd";

var_dump(strcmp($str, "bEg", 3));
var_dump(strncasecmp($str, "beg", 3));````

laruence commented:
and how does str_begins and str_ends better than substr_compare?
````php
<?php
$str = "bEginningMiddleEnd";

var_dump(substr_compare($str, "End", -3));
var_dump(substr_compare($str, "end", -3, 3, true));
?>
````

ju1ius commented:
> how does str_begins different from strncmp/strncasecmp?
> and how does str_begins and str_ends better than substr_compare?

Without getting into micro-optimization,
```php
str_ends($str, $end);
```
is easier to write must more importantly _a lot easier to read_ than
```php
substr_compare($str, $end, -strlen($end), strlen($end), true);
```
Just from the name of it anybody should understand what it does without having to check the documentation.

It would have been even better to name them `str_endswith()` and `str_startswith()`, which read like: «Does this string starts with this other string?», whereas `str_ends` sounds more like «Does this string ends?».
My $0.2

staabm commented:
IMO this can easily achieved in userland with a simple function.

ju1ius commented:
> IMO this can easily be achieved in userland with a simple function.

@staabm  Of course !

But this is a so common task that most programming languages have their native implementations:
* JS: [String.prototype.startsWith](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/String/startsWith)
* Python [str.startswith](https://docs.python.org/2/library/stdtypes.html#str.startswith)
* Ruby [String.start_with?](http://ruby-doc.org/core-2.1.1/String.html#method-i-start_with-3F)
* Go [HasPrefix](https://golang.org/pkg/strings/#HasPrefix)
* Swift [String.hasPrefix](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html)
* Haskell [startswith](https://hackage.haskell.org/package/MissingH-1.4.0.1/docs/Data-String-Utils.html#v:startswith)
* Why not PHP? :wink:

cmb69 commented:
> It would have been even better to name them str_endswith() and str_startswith(), […]

I would fully agree, if these were methods of a string object (`$str->startsWith('foo')` reads so fine), but it appears to be confusing for functions: `str_startswith($str, 'foo')` or `str_startwith('foo', $str)` :-/

nikic commented:
Not sure I see the "functions" issue. "startswith" still imposes an ordering, that the first arg startswith the second arg. Unlike just "starts" which would imply the reverse argument order (first arg starts second arg).

cmb69 commented:
@nikic [Haskell](https://hackage.haskell.org/package/MissingH-1.4.0.1/docs/Data-String-Utils.html#v:startswith) appears to differ. :-)

nikic commented:
Heh. It kinda does make sense to invert the argument order in Haskell in particular, because this makes partial application more useful...

cmb69 commented:
Yes, that makes particular sense for Haskell (and other functional languages), but they could have called it `starts`then.

brzuchal commented:
There are already such implementations based on @nikic extension in here https://github.com/rossriley/php-scalar-objects#startswith scalar objects in PHP would be awesome feature, would love to have it in core. Great job on this extension from @nikic

ju1ius commented:
> but they could have called it starts then.

But they didn't. :wink: 

Wrt the order of arguments, as @nikic said, you're either in an OO-like language (or library, like glib/gobject) and you expect `$this` to be the first argument, or a functional-like language (or library) and you expect the data to be operated on to be the last argument. Both make perfect sense and it's not confusing as long as the api is _consistent_ throughout the library.

Which raises the infamous issue:
```php
function array_filter(array $array, callable $callback): array;
function array_map(callable $callback, array $array): array;
```
deserving a RFC on it's own ! :laughing:

wkhudgins92 commented:
Based off the discussion here and in the mailing list, I'd like to amend the PR and RFC to have case-insensitive functions separated, and also revise the order of arguments to confirm with the rest of the string library. Am I allowed to edit the RFC before voting to reflect some of the issues raised during discussion?

wkhudgins92 commented:
Edited the code and modified the RFC (https://wiki.php.net/rfc/add_str_begin_and_end_functions) to reflect the comments brought up here and in the mailing list. 

I split the functions into separate functions based on case senstivity, switch the order of `$str` and `$seach_value` to be consistent with the other `str_*` functions. I also switched from `_begins()` and `_ends()` to the singual `_begin()` and `_end()`. In addition, the requested changes now pass the Travis CI tests.

ChadSikorra commented:
Will these have `mb_*` equivalents?

BePsvPT commented on ext/standard/string.c:
```diff
@@ -2445,6 +2445,66 @@ PHP_FUNCTION(substr)
+		RETURN_NULL();
+
+	for (i = 0; i < search_value->len; i++) 
+		if (str->val[i] != search_value->val[i]) 
```
When the length of search_value is greater than str, it may cause overflow?

BePsvPT replied:
When the length of search_value is greater than str, maybe it will access out of bounds?

blar replied:
can you change the the interface to str_begin(string $haytstack, $needle): bool and str_ends(string $haystack, string $needle): bool for consistency with the functions with other string function like stristr, strpos, strrchr?

eddy8 commented:
support~

alex-schneider commented:
An optional parameter `int $offset = 0` could be helpful (at least for the `str_ends_with()` function) to be able to check that a string starts and ends with the same character, but not at the same position in the `$haystack` string, e.g. `/` vs. `/path/`.

kelunik commented:
@wkhudgins92 Any plans to move the RFC to a vote?

itsjavi commented:
I would prefer to have `str_starts` and `str_ends` rather than `str_begin` and `str_end`.
The first two are more natural like in any other languages. You won't find str.begin in mostly any other language... They actually seem to be related to some kind of stream buffer.

Also, the parameters should have this name and order: `str_starts($haystack,  $needle)` and `str_ends($haystack,  $needle)` like we have in methods such as `strstr` and `strpos`, for sake of consistency ;)

Summarizing, these are good additions, but we should aim to have an unified string OOP API (also for arrays) in the future, IMO, rather than adding more and more no-oop helpers in the root namespace.

carusogabriel commented:
What is the status here? Really like this RFC :sweat_smile:

wkhudgins92 commented:
@carusogabriel There's been a lot of high quality, but sometimes conflicting, feedback I've received. I'd like to incorporate some of this feedback then put this up to a vote again. Probably be the best way to move this forward.

wkhudgins92 commented:
I merged in master and updated the code to match the modern style of how the interpreter is written.

I added a test case which I believe addresses @BePsvPT concern about overflow.
I modified the function arguments to be consistent with various other string functions, as noted by @itsjavi @blar . 

I want to rename them to str_begins and str_ends as requested by @carusogabriel and @itsjavi . I'd like to include `mb_*` versions as requested by @ChadSikorra but I'm not sure if that fits within the scope of this RFC or should be its own..

krakjoe commented:
Test is failing: https://travis-ci.org/php/php-src/jobs/541061011#L2692

wkhudgins92 commented:
Thanks @krakjoe . I should have the tests fixed later today. . . I'm going to setup a dedicated VM For fixing these tests locally.

cmb69 commented on ext/mbstring/tests/mb_str_begins.phpt:
```diff
@@ -0,0 +1,55 @@
+
+// EUC-JP - With encoding parameter
+mb_internal_encoding('UTF-8') or print("mb_internal_encoding() failed\n");
+var_dump(mb_str_begins($euc_jp, '0123こ', EUC-JP'));
```
There's an apostrophe missing before `EUC-JP` => syntax error.

wkhudgins92 replied:
Thanks @cmb69 !

wkhudgins92 commented:
Now that I have added the mb_* versions of this and the tests are passing, I will update the RFC and move this towards a vote.

tominventisbe commented:
Sorry to be late to the party, but is there a particular reason that something like `str_contains` isn't included? Perhaps it was out of scope?

It would also be one of those common tasks that you now need to use the less readable `strpos` or `mb_strpos` for.

wkhudgins92 commented:
@tominventisbe I think `str_contains` is a great idea, but I feel like it might be beyond the scope of this RFC...maybe if this gets approved I can do an RFC+PR for that?

wkhudgins92 commented:
One of the Travis builds succeeds, one fails. The failing test is `Bug #78185: file cache only no longer works [ext/opcache/tests/bug78185.phpt]`. I don't think my last commits should cause that test to fail. Does anyone have any ideas what is going on?

cmb69 commented:
If you rebase onto current master, this test likely succeeds.

nikic commented on ext/standard/string.c:
```diff
@@ -2301,6 +2301,90 @@ PHP_FUNCTION(substr)
+        RETURN_BOOL(0);
+    }
+
+    for (i = 0; i < needle->len; i++) 
```
`RETURN_BOOL(memcmp(ZSTR_VAL(haystack), ZSTR_VAL(needle), ZSTR_LEN(needle)) == 0);`

Generally these loops should be replaced by `memcmp` calls.

wkhudgins92 replied:
Really good call here! @nikic I replaced the loops for starts_with. The ends_with are a little trickier so I figure I'll wait to do that until the RFC is approved (if it is approved)!

nikic commented on ext/standard/string.c:
```diff
@@ -2301,6 +2301,90 @@ PHP_FUNCTION(substr)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (needle->len > haystack->len) {
+        RETURN_BOOL(0);
```
```suggestion
        RETURN_FALSE;
```

glensc commented on ext/mbstring/mbstring.c:
```diff
@@ -2581,6 +2609,192 @@ PHP_FUNCTION(mb_strripos)
 /* }}} */
 
+/* {{{ proto boolean mb_str_begins(string haystack, string needle [, string encoding])
+   Checks if haystack begins with needle */
```
```suggestion
   Checks if haystack begins with a needle */
```

azjezz commented:
Thanks for working on this :+1: 

> It would have been even better to name them str_endswith() and str_startswith()

i quite agree with this ( or `str_starts_with` / `str_ends_with` instead ? ), but maybe that's just because I'm more familiar with hack standard library. (  [`HH\Lib\Str\starts_with(_ci)`](https://github.com/hhvm/hsl/blob/master/src/str/introspect.php#L259-L287) / [`HH\Lib\Str\ends_with(_ci)`](https://github.com/hhvm/hsl/blob/master/src/str/introspect.php#L96-L138) )

theodorejb commented:
Is this RFC going to vote before the deadline for PHP 7.4 in a few days?

wkhudgins92 commented:
@theodorejb Yesterday marked 2 weeks of discussion, so I plan to move the RFC to a vote.

wkhudgins92 commented:
The RFC related to this has gone to a vote: https://wiki.php.net/rfc/add_str_begin_and_end_functions

ElvenSpellmaker commented:
Does anyone know why the `mb_*` functions are being voted against in the poll? 🤔

nikic commented:
The RFC unfortunately :( :( :( declined, so I'm closing this PR.

ElvenSpellmaker commented:
That's mad... All thanks to the silly 2/3 thing is seems. >50% is a majority, when will people realise that 😂 =|

glensc commented:
the plain string vs mbstring vote difference seems stupid. probably shouldn't even have the opportunity to vote separately.

however, I'm personally looking for Symfony 5 string component:
- https://symfony.com/blog/new-in-symfony-5-0-string-component

simivar commented:
So... is there any possibility to redo the voting and add those functions together with str_contains? It’s the same family. :/

wkhudgins92 commented:
> So... is there any possibility to redo the voting and add those functions together with str_contains? It’s the same family. :/

I am going to re-raise the RFC+PR for only `str_starts_with()` and `str_ends_with()`

glensc commented:
otoh, maybe instead of adding two sets of functions, add support for substring operators on the language level, like in Python (and Ruby?) do:

- `$string[-2]` = substring last two elements
- `$string[:2]` = substring first two elements
- `$string[2:4]` = substring from offsets 2 to 4

but this can't be done until the $string can be declared as ByteArray or UnicodeString, or maybe add also object support to have custom slicers support like `ArrayAcess` interface exists to emulate arrays on onjects.

guilliamxavier commented:
@glensc: The "slice" syntactic sugar idea (which I have already seen proposed elsewhere BTW) is IMHO orthogonal to thoses functions (which are *not* implemented as `$needle === substr($haystack, ...)`)
